src/workerd/io/actor-cache.c++

```diff
@@ -2156,19 +2156,19 @@ void ActorCache::ensureFlushScheduled(const WriteOptions& options) {
     if (options.allowUnconfirmed) {
       // Don't apply output gate. But, if an exception is thrown, we still want to break the gate,
       // so arrange for that.
-      flushPromise = flushPromise.catch_([this](kj::Exception&& e) {
-        return gate.lockWhile(kj::Promise<void>(kj::mv(e)));
+      flushPromise = flushPromise.catch_([this, opInfo = kj::mv(operationInfo)](kj::Exception&& e) {
+        return gate.lockWhile(kj::Promise<void>(kj::mv(e)), kj::mv(opInfo));
```

Comment:
If additional ops get scheduled later, wouldn't they need their own lockWhile? I'd think that each op needs its own lockwhile, otherwise I don't really understand how we ensure that we're waiting on every operation that got scheduled, rather than just on the first and releasing the output gate when the first op is completed?

Reply:
I thought that that the else branch of this if statement is what would put in place a lockWhile for subsequent ops while a flush is already scheduled: https://github.com/cloudflare/workerd/pull/2056/files#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2171

Reply:
ActorCache naturally batches writes together. One possible (and common!) order of events:

1. User calls put(), which calls ensureFlushScheduled(), which does the `flushPromise = lastFlush.addBranch().attach(...).then([]() { flushImpl() })` call above, scheduling a flush for later. We then `gate.lockWhile()` that `flushPromise` (at least in the branch a couple linesbelow, I guess I put this comment on the wrong line)
2. That flush doesn't happen yet, either because `lastFlush()` hasn't finished yet or because the user's code keeps synchronously running.
3. The user does more operations (setAlarm(), delete(), txn(), or whatever). Each time `ensureFlushScheduled` is called on their behalf, nothing is done because `flushScheduled` and `flushScheduledWithOutputGate` are already true.
4. The `lastFlush` promise finally resolves, so `flushImpl` is called.
5. `flushImpl` grabs all dirty keys / alarms for writing and flushes them together.
6. The flush suceeds, releasing the output gate.

Reply:
Hmm, the !flushScheduledWithOutputGate probably prevents that from functioning how I'm describing.

Reply:
Maybe it would be better to have some `kj::Maybe<kj::Exception>` that we populate inside `ActorCache::flushImplUsingTxn()` (and perhaps the other functions we call when we go to flush different dirty entries) if any of our individual RPCs to storage throw, then when we're in the catch branch of `flushImpl()` [here](https://github.com/cloudflare/workerd/blob/main/src/workerd/io/actor-cache.c%2B%2B#L2630) we can `KJ_LOG` the exception before giving the user the error?

Reply:
Hmm, I see, I think what would be most useful to log here would be

- what flush we're waiting on (since it seems like multiple flushes can be enqueued at once)
- what's in the flush we're waiting on

I'll see if it's feasible to log that info instead.

Reply:
@MellowYarker that sounds interesting too. I'll have a look at putting up a PR for that. Not sure what catch branch you're referring to though, don't immediately see a catch there?

Reply:
Ah, it's the errorhandler of that .then()